### DISCUSSION TITLE
doesnt detect newer python than 3.10

### DISCUSSION THREAD

**Discussion by iocc-git** [blueman-project/blueman#2590] [Q&A]
blueman: 2.4.3
BlueZ: ?
Distribution: almalinux 9
Desktop environment: XFCE

$ cd src
$ git clone https://github.com/blueman-project/blueman.git
$ cd blueman
$ ./autogen.sh
(...)
checking for a Python interpreter with version >= 3.10... none
configure: error: no suitable Python interpreter found
$

[root@nixpix blueman]# dnf -y install python3.12 python3.11
Last metadata expiration check: 0:23:46 ago on Sun 29 Dec 2024 20:47:50 CET.
Package python3.12-3.12.5-2.el9_5.2.x86_64 is already installed.
Package python3.11-3.11.9-7.el9_5.2.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
[root@nixpix blueman]# 

python3.10 does not exist and both 3.11 and 3.12 is a higher version then 3.10

**Comment by cschramm** [comment]
`pkg-config` is used to find dependencies. To provide the necessary data, you'll probably need an RPM package like `python3-devel` or similar.

**Reply by iocc-git** [reply]
I have Package python3-devel-3.9.21-1.el9_5.x86_64 already installed.
python version 3.10 seems to be a very odd choice to require.
The docs here mention 3.9, 3.11 and 3.12.... But not 3.10:
https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/installing_and_using_dynamic_programming_languages/assembly_installing-and-using-python_installing-and-using-dynamic-programming-languages#proc_installing-additional-python-3-tools-for-developers_assembly_installing-and-using-python

And >= means that 3.11 should work, and it doesnt.

**Reply by cschramm** [reply]
3.9 is not supported. 3.11 and 3.12 are. My guess for a package that provides what's required would be e.g. `python3.11-devel`.

Getting to something usable will be very tough if providing Python is already challenging 😅

**Comment by infirit** [comment]
I highly recommend not building from main unless you are willing to suffer the pain of RHEL9. These instructions should still be valid for 2.4.x https://github.com/blueman-project/blueman/issues/2234#issuecomment-1874505236. You'll want to get the latest spec file from fedora and add the patch modification.

**Reply by iocc-git** [reply]
Thanks, I might have been tired and not thinking of that. However, it still doesnt work:

(...)
checking for a Python interpreter with version >= 3.10... none
configure: error: no suitable Python interpreter found
[root@nixpix blueman]# rpm -q python3.12 python3.12-devel python3.11 python3.11-devel
python3.12-3.12.5-2.el9_5.2.x86_64
python3.12-devel-3.12.5-2.el9_5.2.x86_64
python3.11-3.11.9-7.el9_5.2.x86_64
python3.11-devel-3.11.9-7.el9_5.2.x86_64
[root@nixpix blueman]#

I didnt have a choice as building it from source was the only option to get it working. No rpm existed.
I even made an ansible playbook of it and it worked well on many machines. Until now.
No, I can think of more fun stuff to do 🙂

**Reply by cschramm** [reply]
`main` recently raised the minimum Python version to 3.10. It seems like you've successfully used 3.9 before that then.

The build system usually has pkgconfig data like `/usr/lib/pkgconfig/python-3.13.pc` and `/usr/lib/pkgconfig/python3.pc`. Not sure if automake might pick up the versioned ones. From the version of your `python3-devel` package, I expect your `python3.pc` to be 3.9 and that one is not supported. Did you try simply uninstalling that package? It could be that the versioned ones put a `python3.pc` in place as well or otherwise automake might pick up the versioned configs.

Otherwise, I have no idea what concept RHEL has to provide 3.11 or 3.12. In the end, you can always do manual setup like a simple `ln -s python-3.12.pc python3.pc` of course.

**Reply by infirit** [reply]
> checking for a Python interpreter with version >= 3.10... none

You did **not** follow the instruction to build a proper RPM you can install. I don't have a RHEL9 vm setup so you'll have to make do with the attached untested but updated zip. The steps to build the RPM are the same as https://github.com/blueman-project/blueman/issues/2234#issuecomment-1874505236

[blueman_rhel9.zip](https://github.com/user-attachments/files/18278737/blueman_rhel9.zip)

**Reply by cschramm** [reply]
@infirit: Couldn't `BuildRequires:  pkgconfig(python3)` lead to a Python 3.9 getting provided, so that the build won't work, depending on what's in the system's archives?

**Reply by infirit** [reply]
It does and that's intended. The spec file is for 2.4.3 which still supports >=3.8. The patch downgrades the automake requirement so it can be build on rhel9. 

Even if you can convince AM_PATH_PYTHON to find the newer Python version I highly doubt the dependencies blueman needs, like pygobject, are provided in the repos build for the newer version.

But I may missed something so please correct me :-)

**Reply by cschramm** [reply]
Ah, the spec is not about git `main`, sure. Got it.

**Reply by infirit** [reply]
Ah, yeah sorry should have mentioned that.

@iocc-git there is pretty much no way for you to get main build/running on RHEL9, it's just too old. But blueman 2.4.3 should still build and run properly.

**Reply by iocc-git** [reply]
> `main` recently raised the minimum Python version to 3.10. It seems like you've successfully used 3.9 before that then.

Yes it worked fine.

> The build system usually has pkgconfig data like `/usr/lib/pkgconfig/python-3.13.pc` and `/usr/lib/pkgconfig/python3.pc`. Not sure if automake might pick up the versioned ones. From the version of your `python3-devel` package, I expect your `python3.pc` to be 3.9 and that one is not supported. Did you try simply uninstalling that package? It could be that the versioned ones put a `python3.pc` in place as well or otherwise automake might pick up the versioned configs.
> 
> Otherwise, I have no idea what concept RHEL has to provide 3.11 or 3.12. In the end, you can always do manual setup like a simple `ln -s python-3.12.pc python3.pc` of course.

Found them in /usr/lib64/pkgconfig.
Tried linking python-3.10.pc to python-3.11.pc
Tried copy python-3.11.pc to python-3.10.pc and then edit the file so it says 3.10 but points to 3.11
Tried to uninstall python3-devel

No luck, still same error.

**Reply by iocc-git** [reply]
> > checking for a Python interpreter with version >= 3.10... none
> 
> You did **not** follow the instruction to build a proper RPM you can install. I don't have a RHEL9 vm setup so you'll have to make do with the attached untested but updated zip. The steps to build the RPM are the same as [#2234 (comment)](https://github.com/blueman-project/blueman/issues/2234#issuecomment-1874505236)
> 
> [blueman_rhel9.zip](https://github.com/user-attachments/files/18278737/blueman_rhel9.zip)

I had no intention to build a rpm, it looks like it can take some time.
To compile blueman from source is fine for me.

**Reply by cschramm** [reply]
> Tried linking python-3.10.pc to python-3.11.pc Tried copy python-3.11.pc to python-3.10.pc and then edit the file so it says 3.10 but points to 3.11 Tried to uninstall python3-devel
> 
> No luck, still same error.

Linking python3.pc to python-3.11.pc or python-3.12.pc could work. I would expect some better way to provide a python3.pc for the 3.11 and 3.12 devel packages, but I have no clue how things are laid out in RHEL. In any case, trying to provide a python-3.10.pc is pointless. It's nothing better than the 3.11 or 3.12 variants.

**Reply by iocc-git** [reply]
> Ah, yeah sorry should have mentioned that.
> 
> @iocc-git there is pretty much no way for you to get main build/running on RHEL9, it's just too old. But blueman 2.4.3 should still build and run properly.

It did. Thanks.
Before I had 2.4.0 on my main workstation but it had some bugs that now are solved in 2.4.3. Lovely 🙂

**Reply by iocc-git** [reply]
> > Tried linking python-3.10.pc to python-3.11.pc Tried copy python-3.11.pc to python-3.10.pc and then edit the file so it says 3.10 but points to 3.11 Tried to uninstall python3-devel
> > No luck, still same error.
> 
> Linking python3.pc to python-3.11.pc or python-3.12.pc could work. I would expect some better way to provide a python3.pc for the 3.11 and 3.12 devel packages, but I have no clue how things are laid out in RHEL. In any case, trying to provide a python-3.10.pc is pointless. It's nothing better than the 3.11 or 3.12 variants.

as blueman-2.4.3 works fine to compile Im happy with that.
